Creation of Offline Installation topic. (#350)
Mirantis/mke-docs: content/docs/getting-started/offline-installation.md

---
title: Offline installation
weight: 3
---

The installation and upgrade procedures for MKE 4k reflect those of the online
scenario. While the online installation, though, typically uses
`registry.mirantis.com` as the primary OCI registry as the repository for MKE
 4k materials, with the offline scenario you instead specify a private
registry from which to pull the MKE 4k artifacts, images, and charts.

{{< callout type="info" >}}

You can download the MKE 4 artifacts from the releases page.

{{< /callout >}}

## Dependencies ##

- skopeo 1.6.1 or later
- An OCI-based private registry that is accessible from all cluster nodes.
- All MKE 4 artifacts and images must be publicly accessible, with no required authentication.
- The registry must use HTTPS, and the TLS certificates of the registry server

> must be signed by a publicly trusted Certificate Authority.

Reviewer: you removed this sentence from my initial proposal

> Private certificate authorities or self-signed certificates are not currently supported.

but I think it's worth including because some people may not even reailize their private CA isn't publicly trusted

> - The registry must support multi-level nesting, for image names

Reviewer: In the issue, I had an example of what I mean here, and the reason I added it was because "multi-level nesting" is not a commonly used term. Not everyone will realize that it means so I want to have an example just to demonstrate what kind of nesting we require.

> e.g. registry.com/level-one/level-two/level-three/image-name:latest. Some registries only allow one level of nesting, registry.com/level-one/image:latest. Make sure your registry supports deeper nesting for image names.

## Preparation ##

> 1. Download the offline bundle from the releases page:

Reviewer: It's either from the release page, then add a link - https://github.com/MirantisContainers/mke-release/releases
Or using the terminal command below. Whatever is more convenient

   ```bash

>    wget https://packages.mirantis.com/caas/mke_bundle_v4.1.0.tar.gz -O mke_bundle_v4.1.0.tar.gz

Reviewer: I realized that wget isn't available on all platforms by default, so let's change it to `curl`

```suggestion
   curl -L https://packages.mirantis.com/caas/mke_bundle_v4.1.0.tar.gz -o mke_bundle_v4.1.0.tar.gz
```

   ```

2. Transfer the bundle file to a machine that can access your private registry.

3. On the machine with registry access, set the environment variables:

   ```bash

>    export REGISTRY_ADDRESS="<registry_address>"  # Registry hostname and optionally port, e.g. "private-registry.example.com:8080". Must NOT end with a slash '/'
>    export REGISTRY_PROJECT_PATH="<registry-path>"  # Path to the registry project that will store all MKE 4 artifacts. Must NOT end with a slash '/'. E.g. "mke". Registry address and path should make the full registry path. With the examples above, the full path will be REGISTRY_ADDRESS + "/" + REGISTRY_PROJECT_PATH == "private-registry.example.com:8080/mke"
>    export REGISTRY_USERNAME="<username>"  # Username of the account allowed to push
>    export REGISTRY_PASSWORD="<password>"  # Password for the same account
>    export BUNDLE_NAME="mke_bundle_v4.1.0-rc.1.tar.gz"  # The name of previously downloaded bundle file. The file must be located in the same directory where you run the preparation steps

Reviewer: Let's replace double quotes with single quotes. It's safer.
Also, I see my comments remained unchanged, and I am happy to think it's because they're perfect. But in case you didn't want to mess up the script, the comments that are after `#` have no effect and you can modify them as you want to make them more readable.

We can even consider another model where the comments for each variable will be outside of the bash script if that makes sense 

```suggestion
   export REGISTRY_ADDRESS='<registry_address>'  # Registry hostname and optionally port, e.g. "private-registry.example.com:8080". Must NOT end with a slash '/'
   export REGISTRY_PROJECT_PATH='<registry-path>'  # Path to the registry project that will store all MKE 4 artifacts. Must NOT end with a slash '/'. E.g. "mke". Registry address and path should make the full registry path. With the examples above, the full path will be REGISTRY_ADDRESS + "/" + REGISTRY_PROJECT_PATH == "private-registry.example.com:8080/mke"
   export REGISTRY_USERNAME='<username>'  # Username of the account allowed to push
   export REGISTRY_PASSWORD='<password>'  # Password for the same account
   export BUNDLE_NAME='mke_bundle_v4.1.0-rc.1.tar.gz'  # The name of previously downloaded bundle file. The file must be located in the same directory where you run the preparation steps
```

   |------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
   | `.spec.registries.imageRegistry.URL` | Sets your registry address with a project path that contains your MKE 4 images. For example, `private-registry.example.com:8080/mke`. <br><br>The setting must not end with a slash `/`.<br><br>The port is optional.                                                                                                                                                                                                                                                                   |

>    | `.spec.registries.chartRegistry.URL` | Sets your registry address with a project path that contains your MKE 4 helm charts in OCI format. For example, `oci://private-registry.example.com:8080/mke`.<br><br>The setting must always start with  `oci://`, and it must not end with a slash `/` .<br><br>If you uploaded the bundle as previously described, the registry address and path will be the same for chart and image registry, with the only difference being the   `oci://` prefix in the chart registry URL. |

Reviewer: ```suggestion
   | `.spec.registries.chartRegistry.URL` | Sets your registry address with a project path that contains your MKE 4 helm charts in OCI format. For example, `oci://private-registry.example.com:8080/mke`.<br><br>The setting must always start with `oci://`, and it must not end with a slash `/` .<br><br>If you uploaded the bundle as previously described, the registry address and path will be the same for chart and image registry, with the only difference being the `oci://` prefix in the chart registry URL. |
```

>    | `.spec.airgap.enabled = true`        | Indicates that your environment is airgapped. With this setting in place, run proceed with the installation normally by running                                                                                                                                                                                                                                                                                                                         |

Reviewer: ```suggestion
   | `.spec.airgap.enabled = true`        | Indicates that your environment is airgapped.                                                                                                                                                                                                                                                                                                                   |
```

Looks like you forgot to remove that. It was unfinished and out of place

3. Run the `mkectl apply` command.

## Upgrade ##

To perform an offline upgrade from MKE 3 to MKE 4k, you adhere to the
procedure as described in [Migrate from MKE 3.x](../../migrate-from-mke-3),
adding the following flags to the `mkectl upgrade` command:

* `--image-registry=<registry_full_path>`
* `--chart-registry=oci://<registry_full_path>`
* `--mke3-airgapped=true`

{{< callout type="info" >}}

> For detail on the `mkectl upgrade` command flags, refer to the mke4.yml

Reviewer: ```suggestion
For detail on the `mkectl upgrade` command flags, refer to the `mke4.yaml`
```

---
title: Offline installation
weight: 3
---

The installation and upgrade procedures for MKE 4k reflect those of the online

> scenario. While the online installation, though, typically uses

Reviewer 2: ```suggestion
scenario. While the online installation typically uses
```

`registry.mirantis.com` as the primary OCI registry for MKE
 4k materials, with the offline scenario you instead specify a private

> registry from which to pull the MKE 4k artifacts, images, and charts.

Reviewer: ```suggestion
registry from which to pull the MKE 4k artifacts, images and charts.
```

It's two types of artifacts: images and charts.

{{< callout type="info" >}}

You can download the MKE 4 artifacts from the [mke-release GitHub repo](https://github.com/MirantisContainers/mke-release/releases).

{{< /callout >}}

## Dependencies ##

- [skopeo](https://github.com/containers/skopeo) 1.6.1 or later
- An OCI-based private registry that is accessible from all cluster nodes.

>   - All MKE 4 artifacts and images must be publicly accessible, with no required authentication.

Reviewer: ```suggestion
  - All MKE 4 images and charts must be publicly accessible, with no required authentication.
```

  - The registry must use HTTPS, and the TLS certificates of the registry server
  must be signed by a publicly trusted Certificate Authority. Private certificate authorities or self-signed certificates are not currently supported.
  - The registry must support multi-level nesting. For example,
    `registry.com/level-one/level-two/level-three/image-name:latest`. Some
    registries only allow one level of nesting, such as
    `registry.com/level-one/image:latest`, so verify that your registry
    supports deeper nesting for image names.

## Preparation ##

1. Download the offline bundle, either from the [mke-release GitHub
   repo](https://github.com/MirantisContainers/mke-release/releases), or
   from the command line as follows:

     ```bash

>      curl -L https://packages.mirantis.com/caas/mke_bundle_v4.1.0.tar.gz -o mke_bundle_v4.1.0.tar.gz

Reviewer: we don't have versioning of the docs yet, but we will need to update this link with every new release; just writing it to create a mental note

| REGISTRY_USERNAME= '<username>'                  | Username for the account that is allowed to push.                                                                                                                                                                                                                                           |
| REGISTRY_PASSWORD= '<password>'                  | Password for the account that is allowed to push.                                                                                                                                                                                                                                           |
| BUNDLE_NAME= 'mke_bundle_v<mke4-version>.tar.gz' | The name of previously downloaded bundle file, which must be located in the same directory in which you run the preparation steps.                                                                                                                                                          |

4. Upload the MKE 4k images and helm charts to your private registry:

>    ```

Reviewer: ```suggestion
   ```bash
```

to add proper syntax highlight

   # Login to the registry
   skopeo login "$REGISTRY_ADDRESS" -u "$REGISTRY_USERNAME" -p "$REGISTRY_PASSWORD"

   # Extract the bundle
   tar -xzf "$BUNDLE_NAME" -C ./

   # Iterate over bundle artifacts and upload each one using skopeo
   for archive in $(find ./bundle -print | grep ".tar"); do
     # Form the image name from the archive name
     img=$(basename "$archive" | sed 's~\.tar~~' | tr '&' '/' | tr '@' ':');

     echo "Uploading $img";
     # Copy artifact from local oci archive to the registry
     skopeo copy -q --retry-times 3 --multi-arch all "oci-archive:$archive" "docker://$REGISTRY_ADDRESS/$REGISTRY_PROJECT_PATH/$img";
   done;
   ```

## Installation ##

{{< callout type="info" >}}

> For information on performing an upgrade to an existing installation in an

Reviewer: ```suggestion
For information on performing an upgrade to an existing MKE 3 installation in an
```

Reviewer: this only applies to MKE 3, but the current wording suggest it applies to MKE 4 -> 4 upgrade which is not the case

airgap environment, refer to [Offline
upgrade](../../migrate-from-mke-3/#offline-upgrade).

{{< /callout >}}

1. Refer to the [Create a Cluster](../create-cluster/#initialize-deployment) procedure for detail on

> how to create a `mke4.yaml` configuration file.

Reviewer 2: ```suggestion
how to create an `mke4.yaml` configuration file.
```

   | `.spec.registries.chartRegistry.URL` | Sets your registry address with a project path that contains your MKE 4 helm charts in OCI format. For example, `oci://private-registry.example.com:8080/mke`.<br><br>The setting must always start with `oci://`, and it must not end with a slash `/` .<br><br>If you uploaded the bundle as previously described, the registry address and path will be the same for chart and image registry, with the only difference being the `oci://` prefix in the chart registry URL. |
   | `.spec.airgap.enabled = true`        | Indicates that your environment is airgapped.                                                                                                                                                                                                                                                                                                                   |

3. Run the `mkectl apply` command.

## MKE 4 versus MKE 3 ##

MKE 3 requires the use of the `docker load` command to load offline bundles
directly into Docker on every cluster node. While this approach does not
require you to have a private registry, it also means that the cluster cannot

> repull the image should any of the loaded images go missing. As such, MKE 3

Reviewer 2: ```suggestion
re-pull the image should any of the loaded images go missing. As such, MKE 3
```
